pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,268 @@
+		mixedversion.NumUpgrades(expectedMajorUpgrades),
+		mixedversion.DisableSkipVersionUpgrades,
+	)
+
+	destMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(sp.srcNodes+1, sp.srcNodes+sp.dstNodes),
+		mixedversion.MinimumSupportedVersion(minSupportedVersion),
+		mixedversion.AlwaysUseLatestPredecessors,
+		mixedversion.NumUpgrades(expectedMajorUpgrades),
+		mixedversion.DisableSkipVersionUpgrades,
+	)
+	return &c2cMixed{
+		sourceMvt:           sourceMvt,
+		destMvt:             destMvt,
+		sp:                  sp,
+		t:                   t,
+		c:                   c,
+		fingerprintArgsChan: make(chan fingerprintArgs, 1),
+		workloadStarter:     make(chan struct{}),
+		fingerprintChan:     make(chan int64, 1),
+	}
+}
+
+type fingerprintArgs struct {
+	retainedTime hlc.Timestamp
+	cutoverTime  hlc.Timestamp
+}
+
+type c2cMixed struct {
+	sourceMvt, destMvt  *mixedversion.Test
+	sp                  replicationSpec
+	t                   test.Test
+	c                   cluster.Cluster
+	fingerprintArgsChan chan fingerprintArgs
+	fingerprintChan     chan int64
+
+	startTime       hlc.Timestamp
+	ingestionJobID  catpb.JobID
+	workloadStarter chan struct{}
+	workloadStopper mixedversion.StopFunc
+}
+
+func (cm *c2cMixed) SetupHook(ctx context.Context) {
+	pgUrlChan := make(chan *url.URL, 1)
+
+	cm.sourceMvt.OnStartup("source: create app tenant", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		srcNode := cm.c.Node(1)
+		srcClusterSetting := install.MakeClusterSettings()
+
+		deprecatedCreateInMemoryTenant(ctx, cm.t, cm.c, sourceTenantName, cm.c.Range(1, cm.sp.srcNodes), false)
+
+		addr, err := cm.c.ExternalPGUrl(ctx, l, srcNode, roachprod.PGURLOptions{})
+		if err != nil {
+			return err
+		}
+
+		pgURL, err := copyPGCertsAndMakeURL(ctx, cm.t, cm.c, srcNode, srcClusterSetting.PGUrlCertsDir, addr[0])
+		if err != nil {
+			return err
+		}
+		pgUrlChan <- pgURL
+		close(cm.workloadStarter)
+		return nil
+	})
+
+	cm.destMvt.OnStartup("dest: start replication", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		sourcePGURL := <-pgUrlChan
+		return h.System.QueryRow(r, fmt.Sprintf("CREATE TENANT %q FROM REPLICATION OF %q ON '%s'", destTenantName, sourceTenantName, sourcePGURL.String())).Scan(&cm.ingestionJobID)
+	})
+}
+
+func (cm *c2cMixed) WorkloadHook(ctx context.Context) {
+	tpccInitCmd := roachtestutil.NewCommand("./cockroach workload init tpcc").
+		Arg("{pgurl%s:%s}", cm.c.Range(1, cm.sp.srcNodes), sourceTenantName)
+	tpccRunCmd := roachtestutil.NewCommand("./cockroach workload run tpcc").
+		Arg("{pgurl%s:%s}", cm.c.Range(1, cm.sp.srcNodes), sourceTenantName).
+		Option("tolerate-errors").
+		Flag("warehouses", 100)
+		// the init command currently runs on init, but the source tenant needs to be created first :/
+	cm.workloadStopper = cm.sourceMvt.Workload("tpcc", cm.c.Range(1, cm.sp.srcNodes), tpccInitCmd, tpccRunCmd)
```

Comment:
@renatolabs the test is currently flaking when running the tpcc init command which gets called via `OnStartup`, but that could race with the  source's startup hook which creates the tenant I want to begin the workload on. To address, I could:
- wait until you push a way to customize tenant initiation
- somehow run the tpcc init and tpcc run cmds in the same `BackgroundFunc`, though the args of `BackgroundFunc` do not provide `cluster`, required to run a cmd. 

Thoughts?

Reply:
I'd say just wait for that work to finish. I've been experimenting with an approach in a branch [1] and will update you once it's ready. 

[1] https://github.com/renatolabs/cockroach/tree/rc/mixedversion-custom-tenant-creation